epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java

```diff
@@ -0,0 +1,146 @@
+
+  /** Maintain visibility item indexed by view id (identity hashcode) */
+  private final SparseArray<EpoxyVisibilityItem> visibilityIdToItemMap = new SparseArray<>();
+
+  /** listener used to process scroll, layout and attach events */
+  private final Listener listener = new Listener();
+
+  @Nullable
+  private RecyclerView attachedRecyclerView = null;
+
+  /**
+   * Attach the tracker.
+   *
+   * @param recyclerView The RecyclerView.
+   */
+  public void attach(@NonNull RecyclerView recyclerView) {
+    attachedRecyclerView = recyclerView;
+    recyclerView.addOnScrollListener(this.listener);
+    recyclerView.addOnLayoutChangeListener(this.listener);
+    recyclerView.addOnChildAttachStateChangeListener(this.listener);
+  }
+
+  /**
+   * Detach the tracker
+   *
+   * @param recyclerView The RecyclerView.
+   */
+  public void detach(@NonNull RecyclerView recyclerView) {
+    recyclerView.removeOnScrollListener(this.listener);
+    recyclerView.removeOnLayoutChangeListener(this.listener);
+    recyclerView.removeOnChildAttachStateChangeListener(this.listener);
+    attachedRecyclerView = null;
+  }
+
+  private void processChildren() {
+    final RecyclerView recyclerView = attachedRecyclerView;
+    if (recyclerView != null) {
+      for (int i = 0; i < recyclerView.getChildCount(); i++) {
+        final View child = recyclerView.getChildAt(i);
+        if (child != null) {
+          processChild(child);
+        }
+      }
+    }
+  }
+
+  private void processChild(@NonNull View child) {
+    processChild(child, false);
+  }
+
+  private void processChild(@NonNull View child, boolean detachEvent) {
+    final RecyclerView recyclerView = attachedRecyclerView;
+    if (recyclerView != null) {
+      recyclerView.getChildViewHolder(child);
+      final ViewHolder holder = recyclerView.getChildViewHolder(child);
+      if (holder instanceof EpoxyViewHolder) {
+        processVisibilityEvents(recyclerView, (EpoxyViewHolder) holder,
+            recyclerView.getLayoutManager().canScrollVertically(), detachEvent);
+      }
+    }
+  }
+
+  private void processVisibilityEvents(
+      @NonNull RecyclerView recyclerView,
+      @NonNull EpoxyViewHolder epoxyHolder,
+      boolean vertical, boolean detachEvent
+  ) {
+    if (epoxyHolder.getAdapterPosition() == RecyclerView.NO_POSITION) {
+      return;
+    }
+
+    final View itemView = epoxyHolder.itemView;
+    final int id = System.identityHashCode(itemView);
+
+    EpoxyVisibilityItem vi = visibilityIdToItemMap.get(id);
+    if (vi == null) {
+      vi = new EpoxyVisibilityItem();
+      visibilityIdToItemMap.put(id, vi);
+    }
+
+    if (vi.getAdapterPosition() != epoxyHolder.getAdapterPosition()) {
+      // EpoxyVisibilityItem being re-used for a different position
+      vi.reset(epoxyHolder.getAdapterPosition());
+    }
+
+    if (vi.update(itemView, recyclerView, vertical)) {
+      // View is measured, process events
+      vi.handleVisible(epoxyHolder);
```

Comment:
It seems inefficient here to have to call every single one of these handle methods. It looks like it is possible to be both visible and invisible at the same time here (if detach event = true), which feels like a bug. Perhaps you should let detach events "win" and not fire their counterpart (e.g. only fire handleInvisible if the detach event is true).

Reply:
Sorry, to be clearer about my comment, it seems like a possible source of bugs that both visible and invisible could be true if detach = true. It seems like it should be 

```
if(!vi.handleVisible(eh)) { vi.handleInvisible(eh) } 
```

or inverted, since invisible is the only state of those two that takes detach into account. If you don't do this, then the callbacks could emit a sequential visible + invisible event.